hack/deployer/runner/k3d.go

```diff
@@ -0,0 +1,204 @@
+}
+
+func (k *K3dDriver) Execute() error {
+	switch k.plan.Operation {
+	case CreateAction:
+		return k.create()
+	case DeleteAction:
+		return k.delete()
+	}
+	return nil
+}
+
+func (k *K3dDriver) create() error {
+	// Do not have k3d modify the kubeconfig by default, as it ends up owned by the root user.
+	cmd := k.cmd("cluster", "create", "--image", k.plan.K3d.NodeImage, "--kubeconfig-update-default=false", "--kubeconfig-switch-context=false")
+	if cmd == nil {
+		return fmt.Errorf("failed to create k3d cluster")
+	}
+	err := cmd.Run()
+	if err != nil {
+		return err
+	}
+
+	// Get kubeconfig from k3d
+	kubeCfg, err := k.getKubeConfig()
+	if err != nil {
+		return err
+	}
+	defer os.Remove(kubeCfg.Name())
+
+	clusterName := fmt.Sprintf("%s-%s", "k3d", k.plan.ClusterName)
+	userName := fmt.Sprintf("admin@%s", clusterName)
+	// First try to remove any existing kubeconfig entries to ensure a clean merge
+	if err := removeKubeconfig(clusterName, clusterName, userName); err != nil {
+		return err
+	}
+
+	// Call mergeKubeconfig to properly merge the kubeconfig without incorrectly overwriting ownership.
+	if err := mergeKubeconfig(kubeCfg.Name()); err != nil {
+		return err
+	}
+
+	// Activate the kubeconfig context to replicate the behavior of the other runners.
+	if err := activateKubeconfig(clusterName); err != nil {
+		return err
+	}
+
+	tmpStorageClass, err := k.createTmpStorageClass()
+	if err != nil {
+		return err
+	}
+
+	return kubectl("--kubeconfig", kubeCfg.Name(), "apply", "-f", tmpStorageClass)
+}
+
+func (k *K3dDriver) delete() error {
+	// The cluster delete operation will not properly remove the kubeconfig entry
+	// as the /home directory isn't mounted in the k3d flow.
+	cmd := k.cmd("cluster", "delete")
+	if cmd == nil {
+		return fmt.Errorf("failed to create k3d cluster")
+	}
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+	// Manually remove the kubeconfig entry.
+	clusterName := fmt.Sprintf("%s-%s", "k3d", k.plan.ClusterName)
+	userName := fmt.Sprintf("admin@%s", clusterName)
+	// ignore errors when removing clusters/users/contexts using k3d
+	_ = removeKubeconfig(clusterName, clusterName, userName)
+	return nil
+}
+
+func (k *K3dDriver) cmd(args ...string) *exec.Command {
+	params := map[string]interface{}{
+		"ClusterName":    k.plan.ClusterName,
+		"SharedVolume":   env.SharedVolumeName(),
+		"K3dClientImage": k.clientImage,
+		"K3dNodeImage":   k.nodeImage,
+		"Args":           args,
+	}
+
+	dockerSocket := "/var/run/docker.sock"
+
+	var socketExists bool
+	_, err := os.Stat(dockerSocket)
+	socketExists = !os.IsNotExist(err)
+
+	// If we are on macOS and the docker socket does not exist, we need to
+	// fall back to using the docker socket in the user's home directory
+	// as with recent changes for docker desktop /var/run/docker.sock
+	// can be created/used, and the docker socket in $HOME errors.
+	if runtime.GOOS == "darwin" && !socketExists {
+		dockerSocket = "$HOME/.docker/run/docker.sock"
+	}
```

Comment:
This seems to the same logic as in `kind.go`, should we share it in a function?
Something along the lines of:

```golang
const defaultDockerSocket = "/var/run/docker.sock"

func getDockerSocket() (string, error) {
	_, err := os.Stat(defaultDockerSocket)
	if err != nil {
		if os.IsNotExist(err) {
			// If we are on macOS and the docker socket does not exist, fall back
			if runtime.GOOS == "darwin" {
				retun "$HOME/.docker/run/docker.sock",nil
			}
		} else {
			// Any other error (e.g. permission denied) should error out
			return "", err
		}
	}
	return defaultDockerSocket, nil
}
```